token.go

```diff
@@ -566,3 +603,95 @@ func processResponse(sess *tdsSession, ch chan tokenStruct) {
 		}
 	}
 }
+
+func processResponse(ctx context.Context, sess *tdsSession, ch chan tokenStruct) {
+	defer func() {
+		close(ch)
+	}()
+	doneChan := ctx.Done()
+	cancelInProgress := false
+	cancelledByContext := false
+	var cancelError error
+
+	// loop over multiple responses
+	for {
```

Comment:
I don't see anything bad, but I'm having trouble following all the loops and if/elses. I'm going to guess we could clean it up a bit, but it is probably fine for now.